universe/ValueRefs.cpp

```diff
@@ -308,22 +308,24 @@ std::string ValueRefBase::InvariancePattern() const {
         .append(ConstantExpr()                  ? "C" : "c");
 }
 
-MeterType NameToMeter(const std::string& name) {
-    auto it = std::find_if(METER_TO_NAME.begin(), METER_TO_NAME.end(),
+MeterType NameToMeter(const std::string_view name) {
+    auto it = std::find_if(NAME_TO_METER.begin(), NAME_TO_METER.end(),
         [&name](const std::pair<std::string_view, MeterType>& elem) {
             return elem.first == name;
         }
     );
 
-    return (it != METER_TO_NAME.end()) ? it->second : MeterType::INVALID_METER_TYPE;
+    return (it != NAME_TO_METER.end()) ? it->second : MeterType::INVALID_METER_TYPE;
 }
 
 const std::string& MeterToName(const MeterType& meter) {
-    for (auto& [name, type] : METER_TO_NAME) {
-        if (type == meter)
-            return std::string{name}; // CodeQL reports "Returning stack-allocated memory", but the name structured binding should be a reference as far as I can tell
-    }
-    return EMPTY_STRING;
+    auto it = std::find_if(NAME_TO_METER.begin(), NAME_TO_METER.end(),
```

Comment:
Part of the suggestion was to replace this linear search with `find_if` with an index into an array. So instead of storing an `array<pair<string_view, MeterType>>`, you can convert the `MeterType` to an index (accounting for `INVALID_METER_TYPE = -1` by offsetting the value before casting to `std::size_t`) and just use the index to get the name from the `array<string_view>` as `array[index]`.

Reply:
Instead of offsetting I did the following:
```cpp
const std::string& MeterToName(const MeterType meter) {
    return (meter == MeterType::INVALID_METER_TYPE) ? EMPTY_STRING : NAME_BY_METER[static_cast<std::size_t>(meter)];
}
```

Reply:
probably OK in practice, though potentially introduces an unnecessary branch that the offsetting variant could avoid.

Reply:
Went with offsetting in the end.